applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -137,37 +166,112 @@ static void handle_dvfs_error(int32_t err)
 		timeout = DVFS_RETRY_INITIALIZATION_TIMEOUT;
 	} else {
 		LOG_ERR("DVFS freq change returned with error: %d", err);
-		module_set_state(MODULE_STATE_ERROR);
-		module_state = STATE_ERROR;
-		cancel_dvfs_retry_work();
+		dvfs_fatal_error();
 		return;
 	}
 	(void) k_work_reschedule(&dvfs_retry.retry_work, timeout);
 }
 
-static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+static uint32_t check_required_frequency(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(dvfs_freq_array) - 1; i++) {
+		if (dfvs_requests_state_bitmask & dvfs_freq_array[i].bitmask) {
+			return dvfs_freq_array[i].freq;
+		}
+	}
+
+	/* If no state is active, return the lowest frequency. */
+	return dvfs_freq_array[ARRAY_SIZE(dvfs_freq_array) - 1].freq;
+}
+
+static void dvfs_notify_cb(struct onoff_manager *srv,
+			   struct onoff_client *cli,
+			   uint32_t state,
+			   int res)
 {
-	int32_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+	(void) k_work_submit(&dvfs_notify_work);
+}
 
+static void set_dvfs_freq(uint32_t target_freq)
+{
+	int ret;
+
+	if (spec.frequency != 0) {
+		ret = nrf_clock_control_release(dvfs_clock_dev, &spec);
+		if (ret < 0) {
+			LOG_ERR("Failed to release requested clock specs, error: %d", ret);
+			dvfs_fatal_error();
+			return;
+		}
+	}
+
+	sys_notify_init_callback(&cli.notify, dvfs_notify_cb);
+
+	spec.frequency = target_freq;
+	ret = nrf_clock_control_request(dvfs_clock_dev, &spec, &cli);
 	if (ret) {
 		handle_dvfs_error(ret);
 	} else {
-		current_freq = target_freq;
-		LOG_INF("Have requested %s frequency",
-			get_dvfs_frequency_setting_name(target_freq));
-		cancel_dvfs_retry_work();
+		LOG_INF("Have requested %" PRIu32 " frequency", target_freq);
+		requested_freq = target_freq;
+		request_in_progress = true;
 	}
 }
 
-static enum dvfs_frequency_setting check_required_frequency(void)
+static void dvfs_frequency_update(void)
 {
-	if (dfvs_requests_state_bitmask & dvfs_high_freq_bitmask) {
-		return DVFS_FREQ_HIGH;
-	} else if (dfvs_requests_state_bitmask & dvfs_medlow_freq_bitmask) {
-		return DVFS_FREQ_MEDLOW;
-	} else {
-		return DVFS_FREQ_LOW;
+	uint32_t required_freq = check_required_frequency();
+
+	if ((!k_work_delayable_is_pending(&dvfs_retry.retry_work)) &&
+	    !request_in_progress && (required_freq != current_freq)) {
+		set_dvfs_freq(required_freq);
+	} else if ((required_freq == current_freq) &&
+		   (k_work_delayable_is_pending(&dvfs_retry.retry_work))) {
+		(void) k_work_cancel_delayable(&dvfs_retry.retry_work);
```

Comment:
Please set `retries_cnt` to zero here too.

Reply:
We shouldn't do it because of mechanism mentioned in https://github.com/nrfconnect/sdk-nrf/pull/17562#discussion_r1812602167

We could end up in endless loop where we would clear retries_cnt  even if we don't make a retry correctly.

Reply:
Disused offline.